plugin/type_hierarchy.py

```diff
@@ -0,0 +1,166 @@
+from .core.paths import simple_path
+from .core.promise import Promise
+from .core.protocol import TypeHierarchyItem
+from .core.protocol import TypeHierarchyPrepareParams
+from .core.protocol import TypeHierarchySubtypesParams
+from .core.protocol import TypeHierarchySupertypesParams
+from .core.protocol import Request
+from .core.registry import new_tree_view_sheet
+from .core.registry import get_position
+from .core.registry import LspTextCommand
+from .core.registry import LspWindowCommand
+from .core.sessions import Session
+from .core.tree_view import TreeDataProvider
+from .core.tree_view import TreeItem
+from .core.typing import cast
+from .core.typing import IntEnum, List, Optional
+from .core.views import make_command_link
+from .core.views import SYMBOL_KINDS
+from .core.views import text_document_position_params
+from functools import partial
+import sublime
+import weakref
+
+
+class TypeHierarchyDirection(IntEnum):
+    Supertypes = 1
+    Subtypes = 2
+
+
+class TypeHierarchyDataProvider(TreeDataProvider):
+
+    def __init__(
+        self,
+        weaksession: 'weakref.ref[Session]',
+        direction: TypeHierarchyDirection,
+        root_elements: List[TypeHierarchyItem]
+    ) -> None:
+        self.weaksession = weaksession
+        self.direction = direction
+        self.root_elements = root_elements
+        session = self.weaksession()
+        self.session_name = session.config.name if session else None
+
+    def get_children(self, element: Optional[TypeHierarchyItem]) -> Promise[List[TypeHierarchyItem]]:
+        if element is None:
+            return Promise.resolve(self.root_elements)
+        session = self.weaksession()
+        if not session:
+            return Promise.resolve([])
+        if self.direction == TypeHierarchyDirection.Supertypes:
+            params = cast(TypeHierarchySupertypesParams, {'item': element})
+            return session.send_request_task(Request.supertypes(params)).then(self._ensure_list)
+        elif self.direction == TypeHierarchyDirection.Subtypes:
+            params = cast(TypeHierarchySubtypesParams, {'item': element})
+            return session.send_request_task(Request.subtypes(params)).then(self._ensure_list)
```

Comment:
Less code:

```suggestion
        if self.direction == TypeHierarchyDirection.Supertypes:
            return session.send_request_task(Request.supertypes({'item': element})).then(self._ensure_list)
        if self.direction == TypeHierarchyDirection.Subtypes:
            return session.send_request_task(Request.subtypes({'item': element})).then(self._ensure_list)
```

Note that I also prefer "no elif after return"